py/picca/pk1d/postproc_pk1d.py

```diff
@@ -862,12 +818,370 @@ def compute_average_pk_redshift(
+        with Pool(number_worker) as pool:
+            output_cov = np.array(pool.map(func, p1d_los_table))
+
+    del p1d_los_table
+
+    p1d_weights, covariance_weights, p1d_groups = (
+        output_cov[:, 0, :],
+        output_cov[:, 1, :],
+        output_cov[:, 2, :],
+    )
+    return p1d_weights, covariance_weights, p1d_groups
+
+
+def compute_groups(nbins_k, p1d_los):
+    """Compute the P1D groups for one subforest.
+
+    Arguments
+    ---------
+    nbins_k (int):
+    Number of k bins.
+
+    p1d_los (array-like):
+    Table containing all p1d pixels unordered for one subforest
+
+    Return
+    ------
+    p1d_weights_id  (array-like):
+    Weights associated with p1d pixels for one subforest, used in the calculation of covariance.
+
+    covariance_weights_id (array-like):
+    Weights for one subforest used inside the main covariance sum.
+
+    p1d_groups_id (array-like):
+    Individual p1d pixels grouped in the same wavenumber binning for one subforest
+    """
+    p1d_weights_id = np.zeros(nbins_k)
+    covariance_weights_id = np.zeros(nbins_k)
+    p1d_groups_id = np.zeros(nbins_k)
+
+    for ikbin in range(nbins_k):
+        mask_ikbin = p1d_los["k_index"] == ikbin
+        number_in_bins = len(mask_ikbin[mask_ikbin])
+        if number_in_bins != 0:
+            weight = p1d_los["weight"][mask_ikbin][0]
+            p1d_weights_id[ikbin] = weight
+            covariance_weights_id[ikbin] = np.sqrt(weight / number_in_bins)
+            p1d_groups_id[ikbin] = np.nansum(
+                p1d_los["pk"][mask_ikbin] * covariance_weights_id[ikbin]
+            )
+    return p1d_weights_id, covariance_weights_id, p1d_groups_id
+
+
 def compute_cov(
+    weight_method,
+    mean_p1d_table,
+    nbins_k,
+    izbin,
+    p1d_weights,
+    covariance_weights,
+    p1d_groups,
+):
+    """Compute the covariance of a set of 1D power spectra.
+
+    Arguments
+    ---------
+    weight_method: str,
+    Method to weight the data.
+
+    mean_p1d_table (array-like):
+    Table of mean 1D power spectra, with column 'meanPk'.
+
+    nbins_k (int):
+    Number of k bins.
+
+    izbin (int):
+    Current redshift bin being considered.
+
+    p1d_weights  (array-like):
+    Weights associated with p1d pixels for all subforest, used in the calculation of covariance.
+
+    covariance_weights (array-like):
+    Weights for all subforest used inside the main covariance sum.
+
+    p1d_groups (array-like):
+    Individual p1d pixels grouped in the same wavenumber binning for all subforest
+
+    Return
+    ------
+    covariance_array (array-like):
+    Array of covariance coefficients.
+    """
+
+    if len(p1d_groups) == 0:
+        return np.full(nbins_k * nbins_k, np.nan)
+
+    mean_pk = mean_p1d_table["meanPk"][(nbins_k * izbin) : (nbins_k * (izbin + 1))]
+    mean_pk_product = np.outer(mean_pk, mean_pk)
+
+    sum_p1d_weights = np.nansum(p1d_weights, axis=0)
+    weights_sum_product = np.outer(sum_p1d_weights, sum_p1d_weights)
```

Comment:
this is very confusing naming given that below you have `weights_product_sum`

Reply:
Yes but we have indeed a sum of product and a product of sums. I renamed them with the of, it should be a little bit clearer

Reply:
maybe adding what is summed into the name could help, but not sure...